extjson.go

```diff
@@ -0,0 +1,1648 @@
+// Copyright 2020 by David A. Golden. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package jibby
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"math"
+	"regexp"
+	"strconv"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+// Efficient extended JSON detection:
+//
+// The longest extended JSON key is $regularExpression at 18 letters
+// The shortest extended JSON key is $oid at 3 letters.  Any $-prefixed
+// key outside those lengths isn't extended JSON.  We can switch on
+// length to avoid a linear scan against all keys.
+//
+// $oid
+// $code
+// $date
+// $type -- option for legacy $binary
+// $scope
+// $regex -- legacy regular expression
+// $binary
+// $maxKey
+// $minKey
+// $symbol
+// $options -- option for legacy regular expression
+// $dbPointer
+// $numberInt
+// $timestamp
+// $undefined
+// $numberLong
+// $numberDouble
+// $numberDecimal
+// $regularExpression
+
+// handleExtJSON is called from convertObject to potentially replace a JSON
+// object with a non-document BSON value instead.  If it returns (nil, nil), it
+// means that the input is not extended JSON and that no bytes were consumed
+// from the input.  If it succeeds, then the extended JSON object will have been
+// consumed.
+//
+// If/when a type is definitively determined, the `typeBytePos` of
+// `out` will be overwritten the discovered type.
+//
+// This function generally only validates/dispatches to subroutines.  Those
+// functions are responsible for consuming the full extended JSON object,
+// including the closing terminator.
+func (d *Decoder) handleExtJSON(out []byte, typeBytePos int) ([]byte, error) {
+	// Peek ahead for longest possible extjson key plus closing quote.
+	buf, err := d.json.Peek(19)
+	if err != nil {
+		// May have peeked to end of input, so EOF is OK.
+		if err != io.EOF {
+			return nil, err
+		}
+	}
+
+	// Common case: not extended JSON.
+	if len(buf) > 0 && buf[0] != '$' {
+		return nil, nil
+	}
+
+	// Isolate key
+	quotePos := bytes.IndexByte(buf, '"')
+	if quotePos < 0 {
+		// Key is longer than `$regularExpression"`, so not valid extended JSON.
+		return nil, nil
+	}
+	key := buf[0:quotePos]
+
+	// When we find a key, we can write a type byte and discard from the input
+	// buffer the length of that key plus one for the closing quote.  In
+	// ambiguous cases, we can't assign a type or discard, so we defer that
+	// to a corresponding subroutine.
+	switch len(key) {
+	case 4: // $oid
+		if bytes.Equal(key, jsonOID) {
+			overwriteTypeByte(out, typeBytePos, bsonObjectID)
+			_, _ = d.json.Discard(5)
+			return d.convertOID(out)
+		}
+		return nil, nil
+	case 5: // $code $date $type
```

Comment:
It probably won't help for case 6, but when we get up to 3+ branches I have a suspicion that it might be faster.

Reply:
Good point.  I'm going to open an issue for it.  I wonder how clever compilers/CPUs are and whether switch/if-else wind up speculatively executed in parallel to the point where it might not matter.

Reply:
I don't have a high opinion of the go compiler, but what I can tell you is that it does do a binary search if you do a switch over integral types.